Disallow step changes when slab number is negative

moxcodes opened the pull request:
## Proposed changes

There are some edge cases where the `can_change_step_size` check in `AdamsBashforthN` doesn't manage to identify self-start, and the step ends up changing, which causes surprising breaks when evolution starts.
This fix makes the assumption that self-start has negative slab numbers, which is true for the current implementation, but may need to be changed in future.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

moxcodes committed:
Disallow step changes when slab number is negative

moxcodes committed:
Make PlaneWave1D domain nonuniform

This ensures that the step chooser will ask for a differing step sizes, and
identify any problems with different steps across elements.

nilsdeppe commented on src/Time/TimeSteppers/AdamsBashforthN.hpp:
```diff
@@ -759,9 +759,10 @@ bool AdamsBashforthN::can_change_step_size(
-  return history.size() == 0 or
-         (less(history.back(), time_id.step_time()) and
-          std::is_sorted(history.begin(), history.end(), less));
+  return time_id.slab_number() > 0 and
```
Should this be `>=`? I'm not sure when we can start allowing. I also remember there being some discussion on an function to check `is_self_starting` but I seem to have yet again forgotten if that got added :P

nilsdeppe reviewed (dismissed):
I can confirm this fixes the issues I was having with the LTS evolutions. Thank you! :)

moxcodes commented:
I think the `is_self_starting` isn't in `develop` yet, but I could add that free function.
Also, there is some freedom in where this check can be placed, and if it's philosophically better to put it in `ChangeStepSize` or `take_step`, it should be similarly effective and a very quick change.

nilsdeppe commented:
:+1: I cannot comment as to where the right place to put it is, though :)

moxcodes committed:
Add is_self_starting function for determining self start

moxcodes commented:
Factored the check to a helper function. For now, I've left the check in `AdamsBashforthN`, under the rough motivation that if it is checked in the stepper itself it is less likely to be overlooked in future features that want to check whether the step can be changed.
Cheers!

wthrowe commented on src/Time/SelfStart.hpp:
```diff
@@ -0,0 +1,15 @@
+/// Reports whether the `time_id` is during self start
+///
+/// This currently assumes that the `time_id` will be negative if and only if
+/// self-start is in progress. If self start is modified to alter that behavior,
```
You're missing something about slabs somewhere in here.

wthrowe reviewed (changes requested):
We used to completely avoid running this check during self-start by not running the ChangeStepSize action, which is why this problem appeared now.  I agree with this method of fixing it.

moxcodes commented:
Yeah, I mistakenly introduced the need for this check when I moved the step-taking routines into `take_step` in `ComputeTimeDerivatives`, so step choosing occurred even during the self-start action list, and I overlooked that the check in the time stepper wouldn't capture all cases in self start.

I've squashed in the modification of the sentence to "... assumes that the slab number of the `time_id` will be...".
Cheers!
